frontend/dockerfile/dockerfile_test.go

```diff
@@ -36,48 +36,122 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
-func TestIntegration(t *testing.T) {
-	integration.Run(t, []integration.Test{
-		testNoSnapshotLeak,
-		testCmdShell,
-		testGlobalArg,
-		testDockerfileDirs,
-		testDockerfileInvalidCommand,
-		testDockerfileADDFromURL,
-		testDockerfileAddArchive,
-		testDockerfileScratchConfig,
-		testExportedHistory,
-		testExposeExpansion,
-		testUser,
-		testDockerignore,
-		testDockerignoreInvalid,
-		testDockerfileFromGit,
-		testCopyChown,
-		testCopyWildcards,
-		testCopyOverrideFiles,
-		testMultiStageImplicitFrom,
-		testCopyVarSubstitution,
-		testMultiStageCaseInsensitive,
-		testLabels,
-		testCacheImportExport,
-		testReproducibleIDs,
-		testImportExportReproducibleIDs,
-		testNoCache,
-		testDockerfileFromHTTP,
-		testBuiltinArgs,
-		testPullScratch,
-		testSymlinkDestination,
-		testHTTPDockerfile,
-		testNoSnapshotLeak,
-		testCopySymlinks,
-		testContextChangeDirToFile,
-		testPlatformArgsImplicit,
-		testPlatformArgsExplicit,
-		testExportMultiPlatform,
+type Mode string
+
+var (
+	ModeBuiltin       = Mode("builtin")
+	ModeMasterGateway = Mode("master-gateway")
+)
+
+type SolverFn func(context.Context, client.SolveOpt, chan *client.SolveStatus) (*client.SolveResponse, error)
+type frontend struct {
+	t    *testing.T
+	mode Mode
+}
+
+func (f *frontend) modeIs(m Mode) bool {
+	return f.mode == m
+}
+
+func (f *frontend) solver(c *client.Client) SolverFn {
+	switch f.mode {
+	case ModeBuiltin:
+		return func(ctx context.Context, opt client.SolveOpt, statusChan chan *client.SolveStatus) (*client.SolveResponse, error) {
+			opt.Frontend = "dockerfile.v0"
+			return c.Solve(ctx, nil, opt, statusChan)
+		}
+	case ModeMasterGateway:
+		return func(ctx context.Context, opt client.SolveOpt, statusChan chan *client.SolveStatus) (*client.SolveResponse, error) {
+			opt.Frontend = "gateway.v0"
+			if opt.FrontendAttrs == nil {
+				opt.FrontendAttrs = make(map[string]string)
+			}
+			opt.FrontendAttrs["source"] = "tonistiigi/dockerfile:master"
+			return c.Solve(ctx, nil, opt, statusChan)
+		}
+	default:
+		f.t.Fatalf("unknown frontend mode %q", f.mode)
+		return nil // not reached
+	}
+}
+
+func (f *frontend) dfCmdArgs(ctx, dockerfile string) (string, string) {
+	traceFile := filepath.Join(os.TempDir(), "trace"+identity.NewID())
+	cmd := "build --no-progress"
+	switch f.mode {
+	case ModeBuiltin:
+		cmd += " --frontend dockerfile.v0"
+	case ModeMasterGateway:
+		cmd += " --frontend gateway.v0 --frontend-opt=source=tonistiigi/dockerfile"
```

Comment:
This one should have `:master` on it too. This also explained the failure of `TestDockerfileInvalidCommand` so that commit can now be dropped too.